### PR TITLE
New "module_missing" lint to detect missing, renamed, or non-visible modules

### LINKED ISSUE #482: False negative: Removing an empty module should be a breaking change
### Steps to reproduce the bug with the above code

Check out the demo repository at https://github.com/nmathewson/cargo-semver-checks-demo-1 .

Note that in `main`,  lib.rs is empty.

Note that in `v0.1.0`, lib.rs contains an empty public module.

Run `cargo semver-checks --baseline-rev v0.1.0`.

### Actual Behaviour

The output is:

```
[1115]$ cargo semver-checks --baseline-rev v0.1.0
     Cloning v0.1.0
     Parsing csc-demo-1 v0.1.1 (current)
     Parsing csc-demo-1 v0.1.0 (baseline)
    Checking csc-demo-1 v0.1.0 -> v0.1.1 (minor change)
   Completed [   0.004s] 39 checks; 39 passed, 6 unnecessary
```

### Expected Behaviour

There should be an error, since any code that tried to import the empty module from an external crate will be broken by a removal of the empty module.

I think this is plausible in realistic scenarios.  For example, you could make a crate with an (empty) prelude, and tell people that they are expected to say `use mycrate::prelude::*`, even if you don't yet have anything in the prelude.

### Generated System Information

```#### Software version

cargo-semver-checks 0.22.1

#### Operating system

Linux 6.3.8-200.fc38.x86_64

#### Command-line

```bash
/home/nickm/.cargo/bin/cargo-semver-checks semver-checks --bugreport 
```

#### cargo version

```
> cargo -V 
cargo 1.70.0 (ec8a8a0ca 2023-04-25)
```

#### Compile time information

- Profile: release
- Target triple: x86_64-unknown-linux-gnu
- Family: unix
- OS: linux
- Architecture: x86_64
- Pointer width: 64
- Endian: little
- CPU features: fxsr,sse,sse2
- Host: x86_64-unknown-linux-gnu
```

### Build Configuration

_No response_

### Additional Context

I'm happy to work on this if you can give me some pointers of what to look at and where to start.

### PR DESCRIPTION
**Pull request by nmathewson**
Closes #482

I tried to follow your instructions at #482 and at https://github.com/obi1kenobi/trustfall-rustdoc-adapter/pull/261#issuecomment-1712652676 ; how does this look?

### CODE REVIEW

File: src/lints/module_missing.ron [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+                    }
+                }
+            }
+        }
+    }"#,
+    arguments: {
+        "public": "public",
+        "zero": 0,
+    },
+    error_message: "A publicly-visible module cannot be imported by its prior path. A `pub use` may have been removed, or the module may have been renamed, removed, or made non-visible.",
```
Comment (obi1kenobi) [inline src/lints/module_missing.ron:44]:
```suggestion
    error_message: "A publicly-visible module cannot be imported by its prior path. A `pub use` may have been removed, or the module may have been renamed, removed, or made non-public.",
```

File: test_crates/module_missing/new/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+pub mod bb {
+    // Removing this should cause a warning.
+    //pub mod will_remove {
+    //}
+}
+
+// Making this private should cause a warning
+pub(crate) mod will_make_private {
+    mod e {
+   }
```
Comment (obi1kenobi) [inline test_crates/module_missing/new/src/lib.rs:23]: the spacing is a bit off here, try running `cargo fmt` on the module

File: test_crates/module_missing/new/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    // Removing this, but no warning should happen:
+    // it isn't visible.
+    // pub mod b {
+    // }
+}
+
+pub mod bb {
+    // Removing this should cause a warning.
+    //pub mod will_remove {
+    //}
```
Comment (obi1kenobi) [inline test_crates/module_missing/new/src/lib.rs:17]: For readability let's do space after `//` every time.

File: test_crates/module_missing/old/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+mod a {
+}
+
+mod b {
+    pub mod b {
+    }
```
Comment (obi1kenobi) [inline test_crates/module_missing/old/src/lib.rs:6]: Any reason not to just do `{}` on the same line throughout when the module is empty?

**Review by obi1kenobi — COMMENTED** [review]
Nicely done! All looks correct, just a few tiny nitpicks.

**Comment by nmathewson** [comment]
Pushed the suggested changes. Thanks!

File: test_crates/module_missing/new/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    // it isn't visible.
+    // pub mod b {
+    // }
+}
+
+pub mod bb {
+    // Removing this should cause a warning.
+    //
+    // pub mod will_remove {
+    // }
```
Comment (obi1kenobi) [inline test_crates/module_missing/new/src/lib.rs:18]:
```suggestion
// mod a {}

mod b {
    // Removing this, but no warning should happen:
    // it isn't visible.
    // pub mod b {}
}

pub mod bb {
    // Removing this should cause a warning.
    //
    // pub mod will_remove {}
```

**Review by obi1kenobi — COMMENTED** [review]
Just a minor consistency fix applying the formatting to the commented-out code as well.

**Review by obi1kenobi — APPROVED** [review]

**Comment by obi1kenobi** [comment]
Marked for merge as soon as the tests finish, well done!

Are you interested in writing some more lints, or perhaps improving the contributing docs based on which things weren't obvious? No pressure of course, but if you are interested, I'm happy to keep merging :)

**Comment by nmathewson** [comment]
> Are you interested in writing some more lints, or perhaps improving the contributing docs based on which things weren't obvious? No pressure of course, but if you are interested, I'm happy to keep merging :)

FWIW, I do hope to add documentation, cleanups, and maybe a few more lints, in what free time I have... but of course, that free time is not as abundant as I would like. ;)

**Comment by obi1kenobi** [comment]
Of course! Any PRs you're interested in putting together are welcome and appreciated, regardless of timeline. You're in charge of the pace, I just want to make sure you are never blocked on me.

In terms of next lints, #366 has 3 suggestions that already have the necessary schema and just need lints and test cases, so they are the lowest-hanging fruit. If you'd like something more substantial for lints, we could also use help implementing `#[doc(hidden)]` handling. I'm about to be AFK for a couple of weeks (RustConf, then Strange Loop) but I'd be happy to mentor for that when I'm back if you're interested in it.

**Pull request merged**